✅ NYC Yellow Taxi Data Engineering Capstone – Final Submission

TuoYang263 opened the pull request:
# NYC Yellow Taxi Data Engineering Pipeline 🚖

This project is part of the DataTribe Collective Capstone Challenge. It demonstrates a full data pipeline using real-world NYC Yellow Taxi trip records from January 2023.

## ✨ Goal

Ingest → Clean → Transform → Store → Visualize  
Using tools like Pandas, DuckDB, Streamlit, and optionally Airflow.

## 📁 Project Structure

tuo-yang-nyc-taxi/
├── data/                 # Raw parquet data
├── scripts/              # Ingestion & transformation scripts
├── notebooks/       # Exploratory analysis and plotting
├── dags/                # Airflow DAGs (optional)
├── README.md     # This file
├── Makefile / pyproject.toml    # Build / Dependency setup

TuoYang263 committed:
The 1st commit of tuo-yang-nyc-taxi project

chanukyapekala commented:
I cannot find any code changes, i see only the files being renamed! Is that your intention?

TuoYang263 committed:
tuo yang commit on 2025-05-20 (updates of notebooks)

chanukyapekala reviewed (changes requested):
I cannot see your changes, your file names are changed! Pls commit your code and then open a PR!

eveliinahampus commented:
Hi👋 It seems that you opened the PR quite early. No worries, but if there is any problems with git, github, understanding how forks and pull requests work, let us know! I'll be happy to make tutorial if needed. ☺️

TuoYang263 committed:
commit on 2025-05-21

TuoYang263 committed:
commit on 2025-05-21

TuoYang263 commented:
Hi @chanukyapekala @eveliinahampus! 👋

Apologies for opening the PR too early last time — I totally get how that can be frustrating 😅

At this stage, I've only completed the initial data exploration notebook (`01_explore_raw_data.ipynb`) for the NYC Yellow Taxi dataset. The rest of the pipeline is still in progress — I submitted this early so I can get early feedback on structure, data loading, and any potential red flags before building the full flow.

Here's a quick summary of what I'm planning next:

1. **Transformation Logic with DuckDB**
   - I’ll create a summary table (`trip_summary_hourly`) using SQL in DuckDB.
   - This will include aggregated metrics like trip count, avg distance, avg fare, and avg duration — grouped by hour.
   - I’ll store the result as a `.csv` or `.parquet` file for visualization.

2. **ETL Structure**
   - Python scripts under `scripts/` will be used to:
     - Ingest raw parquet files
     - Transform data into clean analysis tables
     - Optionally upload processed data to BigQuery
   - All scripts will follow a modular and reproducible structure.

3. **Visualization**
   - A separate notebook will be created to visualize the transformed output.
   - I’ll use Seaborn / Matplotlib initially, and possibly a Streamlit app if time allows.

4. **README & Project Docs**
   - I will write a clear README for the project folder, with context, structure, and instructions.

I know this PR is incomplete — but I wanted to show what’s in progress and align with expectations early.

Thanks again for your patience, and I’ll update this PR as soon as the next part is ready! 🙏

P.S. Appreciate the early comments — they helped me catch the PR issue quickly 😅 Still learning the ropes when it comes to pull requests and collaborative workflows, so thanks again for your support!

TuoYang263 committed:
updates of notebooks: 01_explore_raw_data.ipynb

TuoYang263 committed:
commit on 2025-05-22

chanukyapekala commented:
Pls complete the PR from your point of view and then let us know here! It is not a good practice to review in progress PRs, reviewer doesnt understand the state of the project in those cases. So, in your vision, if most of the components are done and is at final stages, it would be then okay for the PR to be viewed but not while you are actually working on the source code.

TuoYang263 commented:
Hi @chanukyapekala, thank you for the reminder — I completely understand your point now.

To avoid triggering unnecessary review notifications, I’ll switch to working mostly locally and only push when a functional unit (e.g., transformation, visualization, docs) is complete. I often commit frequently during development — especially for intermediate saves or debugging checkpoints — but I’ll now keep those commits local until the overall component is ready.

Going forward:
- I’ll develop incrementally on my local branch
- Only push once a section is done and stable
- I’ll use this PR as a staging area until the full flow is ready (EDA → transform → visualization)
- I’ll comment again when it’s finalized and ready for proper review

Thanks again for your guidance — this is helping me align better with good collaboration practices!

chanukyapekala commented:
Hi @TuoYang263 - What is the status of this PR?

TuoYang263 committed:
Merge branch 'datatribe-collective-labs:main' into tuo-yang-capstone-draft

TuoYang263 committed:
Add NYC Yellow Taxi Capstone Project for final review

chanukyapekala commented:
@Akomal : Can you please review this?

Akomal commented on tuo-yang-nyc-taxi/airflow_pipeline/sql/bq/create_trip_summary_hourly.sql:
Just an advice to use full expression in group by instead of alias. Alias works with modern DBs but for production code and from reuse-ability perspective its better to include full expression.

Akomal commented on tuo-yang-nyc-taxi/airflow_pipeline/scripts/load_nyc_yellow_taxi_data.py:
Use logging instead of print statement.

Akomal commented on tuo-yang-nyc-taxi/airflow_pipeline/scripts/transform_trip_data_spark.py:
You can validate schema before transformation and log errors in case schema checks fail.

Akomal commented on tuo-yang-nyc-taxi/airflow_pipeline/scripts/transform_trip_data_spark.py:
```diff
@@ -0,0 +1,141 @@
+from typing import List
+from pathlib import Path
+from scripts import config
+
```
You can validate schema before transformation and log errors in case schema checks fail.

Akomal commented on tuo-yang-nyc-taxi/airflow_pipeline/scripts/load_nyc_yellow_taxi_data.py:
```diff
@@ -0,0 +1,52 @@
+                with open(local_path, "wb") as f:
+                    f.write(response.content)
+            except Exception as e:
+                print(f"❌ Failed to download {url}: {e}")
```
Use logging instead of print statement.

TuoYang263 replied:
Thanks! Updated to use logger.error(...) instead of print() in the latest commit. I also made sure that all scripts now consistently use the shared logger instance for structured logging.

Let me know if you'd like any further logging refinements.

Akomal commented on tuo-yang-nyc-taxi/notebooks/04_analyze_trip_summary.ipynb:
Here it's better to use the file that's saved locally and not call the bigquery. It breaks the flow as said airflow is optional and it can run locally. Also If I understand correctly in notebooks we are just doing EDA for the month of january and in bigquery we will load the data for three months.

TuoYang263 replied:
Thanks for pointing that out! I've updated the notebook to read from the local Parquet files instead of querying BigQuery directly. Just to clarify:

1. 01_explore_raw_data.ipynb focuses on January only

2. 03_visualize_summary.ipynb and downstream notebooks use the full 3-month data for aggregated analysis

This way the notebooks stay self-contained and Airflow/BigQuery remains optional. Appreciate the suggestion!

Akomal commented on tuo-yang-nyc-taxi/README.MD:
```diff
@@ -0,0 +1,307 @@
+Feel free to connect or share feedback:
+
+- **GitHub**: [tuo-yang](https://github.com/TuoYang263)  
+- **LinkedIn**: [tuo-yang-linkedin-profile](https://linkedin.com/in/your-profile)  
```
When I opened the linkedin from preview it opens someone else's linkedin account.

TuoYang263 committed:
Finalize NYC Taxi Capstone project for PR review

TuoYang263 committed:
Fix .gitignore to preserve empty data folders

TuoYang263 committed:
Add raw folder with sample parquet files

TuoYang263 commented on tuo-yang-nyc-taxi/airflow_pipeline/dags/nyc_taxi_etl_dag.py:
```diff
@@ -0,0 +1,102 @@
+    'start_date': datetime(2024, 1, 1),
+    'retries': 2,
+    'retry_delay': timedelta(minutes=5),
+    'email': [],
```
Good catch — I’ve fixed the typo in the latest commit. Thanks!

TuoYang263 commented on tuo-yang-nyc-taxi/airflow_pipeline/sql/raw/trip_summary_hourly.sql:
```diff
@@ -0,0 +1,11 @@
+FROM yellow_taxi
+WHERE trip_distance > 0 AND total_amount > 0
+    AND tpep_pickup_datetime BETWEEN '2023-01-01' AND '2023-02-01'
+GROUP BY pickup_hour
```
Thanks for the suggestion! I've updated the query to use the full expression in the GROUP BY clause instead of the alias, as recommended. Appreciate the point about production readability and reuse — totally makes sense.

Let me know if there's anything else I should tweak.

TuoYang263 commented:
Thanks for the thoughtful and detailed review, @Akomal!

All requested changes have been addressed:

- Replaced `print()` with `logger.warning()` in the download script.
- Rewrote the SQL query to use full expressions in the `GROUP BY` clause.
- Added `raw/` folder with sample parquet files for Jan–Mar 2023, and ensured `.gitignore` preserves empty folders.
- Updated notebooks to use only local data:
  - `01_explore_raw_data.ipynb` analyzes January data.
  - `03_visualize_summary.ipynb` now runs locally using 3-month data.
  - Removed outdated `04_analyze_trip_summary.ipynb` which accessed BigQuery.
- Fixed the LinkedIn link in `README.md` that was pointing to a placeholder.

Additionally, I added:
- A `validate_schema()` helper function to log any missing or unexpected fields for schema awareness.
- Type casting logic for critical columns, as using an explicit Spark schema caused errors in some cases.

Please let me know if you spot anything else that needs improvement. Thanks again for your time and great feedback!

TuoYang263 commented:
Hi @chanukyapekala,
All review comments from @Akomal have been addressed, checks are passing, and all conversations are resolved.
Could you please confirm if any additional review is required, or if we can go ahead and merge this PR?
Thanks for your time and support! 🙌
